File: patterns/Creational/Prototype/RealWorldExamples/Document.php
<?php

/*
|--------------------------------------------------------------------------
| Prototype Design Pattern - Example
|--------------------------------------------------------------------------
| Implement the Prototype Design Pattern to create new instances by cloning 
| existing objects, which can avoid costly initialization steps.
|--------------------------------------------------------------------------
| @category  Design Pattern
| @package   Creational/Prototype/RealWorldExamples
| @author    JawherKl
| @version   1.0.0
| @license   MIT License
| @link      https://github.com/JawherKl/design-patterns-in-php
|--------------------------------------------------------------------------
*/

namespace Creational\Prototype\Example;

/**
 * Prototype Interface.
 * Defines the `__clone()` method that all prototypes must implement.
 */
interface Prototype
{
    public function __clone();
}

/**
 * Concrete Prototype: Document
 */
class Document implements Prototype
{
    private $title;

    private $content;

    /**
     * @var \DateTime
     */
    private $createdAt;

    /**
     * @var Author
     */
    private $author;

    public function __construct(string $title, string $content, Author $author)
    {
        $this->title = $title;
        $this->content = $content;
        $this->author = $author;
        $this->createdAt = new \DateTime();
        $this->author->addDocument($this);
    }

    /**
     * Control what data to copy when cloning.
     * Example:
     * - The cloned document will have "Copy of ..." title.
     * - The author reference remains the same.
     * - The creation date is updated to the current date.
     */
    public function __clone()
    {
        $this->title = "Copy of " . $this->title;
        $this->createdAt = new \DateTime();
        // Add the cloned document to the author's list of documents.
        $this->author->addDocument($this);
    }

    public function getDetails()
    {
        return [
            'Title' => $this->title,
            'Content' => $this->content,
            'Author' => $this->author->getName(),
            'CreatedAt' => $this->createdAt->format('Y-m-d H:i:s')
        ];
    }
}

/**
 * The Author class, which has a collection of documents.
 */
class Author
{
    private $name;

    /**
     * @var Document[]
     */
    private $documents = [];

    public function __construct(string $name)
    {
        $this->name = $name;
    }

    public function getName(): string
    {
        return $this->name;
    }

    public function addDocument(Document $document): void
    {
        $this->documents[] = $document;
    }

    public function getDocuments(): array
    {
        return $this->documents;
    }
}

/**
 * Client code that demonstrates the Prototype Design Pattern.
 */
function clientCode()
{
    // Create an author and a document.
    $author = new Author("Jane Doe");
    $originalDocument = new Document("Design Patterns", "Learning the Prototype Pattern.", $author);

    // Clone the original document.
    $clonedDocument = clone $originalDocument;

    // Display details of both documents.
    echo "Original Document:\n";
    print_r($originalDocument->getDetails());

    echo "\nCloned Document:\n";
    print_r($clonedDocument->getDetails());
}

clientCode();

File: patterns/Creational/Prototype/Go/RealWorldExamples/Document.go
/*
|--------------------------------------------------------------------------
| Prototype Design Pattern - Document Example
|--------------------------------------------------------------------------
| Implement the Prototype Design Pattern to create new instances by cloning 
| existing objects, which can avoid costly initialization steps.
|--------------------------------------------------------------------------
| @category  Design Pattern
| @package   Creational/Prototype/RealWorldExamples
| @version   1.0.0
| @license   MIT License
| @link      https://github.com/Milen Denev/design-patterns-in-go
|--------------------------------------------------------------------------
*/

package main

import (
    "fmt"
    "time"
)

// Prototype Interface.
// Defines the Clone method that all prototypes must implement.
type Prototype interface {
    Clone() Prototype
    GetDetails() map[string]string
}

// Concrete Prototype: Document
type Document struct {
    Title     string
    Content   string
    CreatedAt time.Time
    Author    *Author
}

// NewDocument is the constructor for Document.
func NewDocument(title, content string, author *Author) *Document {
    doc := &Document{
        Title:     title,
        Content:   content,
        CreatedAt: time.Now(),
        Author:    author,
    }
    author.AddDocument(doc)
    return doc
}

// Clone provides a copy of the Document with certain modifications.
func (d *Document) Clone() Prototype {
    clone := *d
    clone.Title = "Copy of " + d.Title
    clone.CreatedAt = time.Now()
    d.Author.AddDocument(&clone)
    return &clone
}

// GetDetails returns the details of the Document.
func (d *Document) GetDetails() map[string]string {
    return map[string]string{
        "Title":     d.Title,
        "Content":   d.Content,
        "Author":    d.Author.Name,
        "CreatedAt": d.CreatedAt.Format("2006-01-02 15:04:05"),
    }
}

// The Author class, which has a collection of documents.
type Author struct {
    Name      string
    Documents []*Document
}

// NewAuthor is the constructor for Author.
func NewAuthor(name string) *Author {
    return &Author{
        Name:      name,
        Documents: []*Document{},
    }
}

// AddDocument adds a Document to the Author's list of documents.
func (a *Author) AddDocument(document *Document) {
    a.Documents = append(a.Documents, document)
}

// Client code that demonstrates the Prototype Design Pattern.
func clientCode() {
    // Create an author and a document.
    author := NewAuthor("Jane Doe")
    originalDocument := NewDocument("Design Patterns", "Learning the Prototype Pattern.", author)

    // Clone the original document.
    clonedDocument := originalDocument.Clone()

    // Display details of both documents.
    fmt.Println("Original Document:")
    for k, v := range originalDocument.GetDetails() {
        fmt.Printf("%s: %s\n", k, v)
    }

    fmt.Println("\nCloned Document:")
    for k, v := range clonedDocument.GetDetails() {
        fmt.Printf("%s: %s\n", k, v)
    }
}

func main() {
    clientCode()
}
